src/RefractAPI.cc

```diff
@@ -269,24 +246,89 @@ std::unique_ptr<IElement> HeaderToRefract(const NodeInfo<snowcrash::Header>& hea
     return std::move(element);
 }
 
-std::unique_ptr<IElement> AssetToRefract(
-    const NodeInfo<snowcrash::Asset>& asset, const std::string& contentType, const std::string& metaClass)
+namespace
 {
-    if (asset.node->empty()) {
-        return nullptr;
+    std::unique_ptr<StringElement> make_asset_element( //
+        std::string content,                           //
+        std::string klass,                             //
+        std::string contentType,                       //
+        const mdp::CharactersRangeSet* sourceMap = nullptr)
+    {
+        auto result = from_primitive_t(SerializeKey::Asset, std::move(content));
+
+        if (!klass.empty())
+            result->meta().set(SerializeKey::Classes, //
+                make_element<ArrayElement>(from_primitive(std::move(klass))));
+
+        if (sourceMap && !sourceMap->empty())
+            result->attributes().set(          //
+                SerializeKey::SourceMap,       //
+                SourceMapToRefract(*sourceMap) //
+            );
+
+        if (!contentType.empty())
+            result->attributes().set(SerializeKey::ContentType, //
+                from_primitive(std::move(contentType)));
+
+        return result;
+    }
+}
+
+namespace
+{
+    using MediaType = apib::parser::mediatype::state;
+
+    apib::parser::mediatype::state addSchemaSubtype(apib::parser::mediatype::state m)
+    {
+        if (m.suffix.empty())
+            m.suffix = m.subtype;
+        m.subtype = "schema";
```

Comment:
I’m not sure this is the right approach, while it works for `application/json` as JSON Schema is `application/schema+json` it might not work for all formats. For example XML DTD (Document Type Definition) has the mime type `application/xml-dtd` (https://www.w3.org/2006/02/son-of-3023/draft-murata-kohn-lilley-xml-04.html#anchor3) not `application/schema+xml`.

There might also be other attributes in the `MediaType` that might not make sense on our schemas if we are generating them. In the case of other unknown parameters. Even `charset` might be set as what the server is returning in its response, the charset for our generated JSON Schema or JSON Schema in the asset is going to be unicode regardless (atleast in current serialisation) as that’s already enforced by API Elements parse result being serialised as JSON (as that specifies unicode).

Reply:
I'm also not sure about the approach with adding `schema`. I will restrict the behavior to json-type'd media types.

However, I feel that the media type of an unknown XML schema is as much `application/schema+xml` as it is `application/xml` as it is `application/beer`. It's neither of these, in general. And in the code we can recognize that ambiguity. Wouldn't it be even better to produce `text/plain` or nothing at all in such cases?

Regarding media type parameters; `master` currently attempts propagating all parameters, I felt this PR has a diff large enough already, and this PR therefore just fully propagates any parameters down to their respective Body/Schema sections. Are you suggesting we should not propagate media type parameters on generated schemas? Or bodies? Because that is not current behavior.

Reply:
> I'm also not sure about the approach with adding schema. I will restrict the behavior to json-type'd media types.

:+1:

> However, I feel that the media type of an unknown XML schema is as much application/schema+xml as it is application/xml as it is application/beer. It's neither of these, in general. And in the code we can recognize that ambiguity. Wouldn't it be even better to produce text/plain or nothing at all in such cases?

Perhaps, the assumption that a schema matches the content-type of the body isn't nessecerily correct but seems a close approximation that was done in the past. There could be cases where I am describing the body in a vastly different content type. Perhaps `text/plain` makes sense if we don't know the semantics behind it (or omitting the content-type in that case).

One example can be:

```
+ Request (application/yaml)
    + Body

            key: value

    + Schema

            {"type": "object"}
```

Where I've used JSON Schema to describe a YAML request. Although not sure in practise anyone would do such a thing.

> Are you suggesting we should not propagate media type parameters on generated schemas? Or bodies? Because that is not current behavior.

Looked like it was introduced based on the diff against some of the fixtures such as `test/fixtures/extend/many.json`.